sys/include/vfs.h

```diff
@@ -0,0 +1,453 @@
+ * @ingroup   sys
+ * @brief     Provides an interface for accessing files and directories from different devices and file systems
+ *
+ * This layer is modeled as a mix between POSIX syscalls (e.g. open) and the
+ * Linux VFS layer implementation, with major reductions in the feature set, in
+ * order to fit the resource constrained platforms that RIOT targets.
+ *
+ * @todo VFS layer reference counting for open files and simultaneous access.
+ *
+ * @{
+ * @file
+ * @brief   VFS layer API specifications
+ * @author  Joakim Nohlgård <joakim.nohlgard@eistec.se>
+ */
+
+#ifndef VFS_H_
+#define VFS_H_
+
+#include <stdint.h>
+#include <sys/stat.h> /* for struct stat */
+#include <sys/types.h> /* for off_t etc. */
+
+#include "kernel_types.h"
+#include "atomic.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef VFS_MAX_OPEN_FILES
+/**
+ * @brief Maximum number of simultaneous open files
+ */
+#define VFS_MAX_OPEN_FILES (16)
+#endif
+
+#ifndef VFS_MAX_MOUNTS
+/**
+ * @brief Maximum number of simultaneous mounted file systems
+ */
+#define VFS_MAX_MOUNTS (4)
+#endif
+
+#ifndef VFS_MOUNT_POINT_LEN
+/**
+ * @brief Longest possible mount point name, including trailing null byte
+ */
+/* string       | required length
+ * "/"          | 2
+ * "/mnt"       | 5
+ * "/home"      | 6
+ * "/var/tmp"   | 9
+ * "/mnt/cdrom" | 11
+ */
+#define VFS_MOUNT_POINT_LEN (8)
+#endif
+
+/**
+ * @brief Used with vfs_bind to bind to any available fd number
+ */
+#define VFS_ANY_FD (-1)
+
+/* Forward declarations */
+struct vfs_file_ops;
+/**
+ * @brief struct @c vfs_file_ops typedef
+ */
+typedef struct vfs_file_ops vfs_file_ops_t;
+struct vfs_file_system_ops;
+/**
+ * @brief struct @c vfs_file_system_ops typedef
+ */
+typedef struct vfs_file_system_ops vfs_file_system_ops_t;
+
+/**
+ * @brief A file system driver
+ */
+typedef struct {
+    const vfs_file_ops_t *f_op;         /**< File operations table */
+    const vfs_file_system_ops_t *fs_op; /**< File system operations table */
+} vfs_file_system_t;
+
+/**
+ * @brief A mounted file system
+ */
+typedef struct {
+    const vfs_file_system_t *fs;     /**< The file system driver for the mount point */
+    char mount_point[VFS_MOUNT_POINT_LEN]; /**< Mount point, e.g. "/mnt/cdrom" */
+    atomic_int_t open_files;   /**< Number of currently open files */
+    void *private_data;        /**< File system driver private data, implementation defined */
+} vfs_mount_t;
+
+/**
+ * @brief An open file
+ *
+ * Similar, but not equal, to struct file in Linux
+ */
+typedef struct {
+    const vfs_file_ops_t *f_op;   /**< File operations table */
+    vfs_mount_t *mp;        /**< Pointer to mount table entry */
+    int flags;              /**< File flags */
+    off_t pos;              /**< Current position in the file */
+    kernel_pid_t pid;       /**< PID of the process that opened the file */
+    void *private_data;     /**< File system driver private data, implementation defined */
```

Comment:
I propose we add a private int value as well, either alongside or in an union with the void *private_data.
It would become handy to store a file handle returned by the underlying file system implementation, for instance.